src/redprl/refiner_inductive_spec.fun

```diff
@@ -0,0 +1,360 @@
+                  in
+                    (boundary, goal :: goals)
+                  end)
+                (trySimplify (args, Abt.substVar (arg, x) bx))
+          | trySimplify' (arg::args, Syn.IND_CONSTR_SPEC_FUN (a,x,bx)) =
+              Option.map
+                (fn (boundary, goals) =>
+                  let
+                    val goalsSpec = EqSpec H (constrs, specctx) ((arg, arg), a)
+                  in
+                    (boundary, goalsSpec @ goals)
+                  end)
+                (trySimplify (args, Abt.substVar (arg, x) bx))
+          | trySimplify' (arg::args, Syn.IND_CONSTR_LINE (x,bx)) =
+              (* XXX no sort-checking *)
+              trySimplify (args, Abt.substVar (arg, x) bx)
+        and trySimplify (args, spec) = trySimplify' (args, Syn.out spec)
+      in
+        trySimplify (args, ConstrDict.lookup constrs label)
+      end
+
+    and simplifyConstr H (constrs, specctx) =
+      fn Syn.IND_SPEC_INTRO args => simplifyIntro H (constrs, specctx) args
+       | Syn.IND_SPEC_FCOM args => simplifyFComTube H (constrs, specctx) args
+
+    fun EqSpecInterBoundary H (constrs, specctx) boundaries =
+      let
+        val goalsOnDiag = List.concat @@
+          List.map
+            (fn (eq, b) =>
+              restrictedEqSpec [eq] H (constrs, specctx)
+                ((b, b), IND_SPECTYPE_SELF))
+            boundaries
+
+        val goalsNotOnDiag = List.concat @@
+          ComKit.enumInterExceptDiag
+            (fn ((eq0, b0), (eq1, b1)) =>
+              restrictedEqSpecIfDifferent [eq0, eq1] H (constrs, specctx)
+                ((b0, b1), IND_SPECTYPE_SELF))
+            (boundaries, boundaries)
+      in
+        goalsOnDiag @ goalsNotOnDiag
+      end
+
+    (* Is it okay to move dimensions upfront? It is banned in Part IV,
+     * and the parser disallows it, but the checker here allows this.
+     *)
+    fun checkConstr' (H, dimset) (constrs, specctx) constr level =
+      case Syn.out constr of
+         Syn.IND_CONSTR_DISCRETE [] => [] (* XXX more refined criterion *)
+       | Syn.IND_CONSTR_KAN boundaries =>
+           let
+             val eqs = List.map #1 boundaries
+             fun inSet dim =
+               case Syn.out dim of
+                  Syn.DIM0 => true
+                | Syn.DIM1 => true
+                | Syn.VAR (v, _) => DimSet.member dimset v
+             val true = List.all (fn (r0, r1) => inSet r0 andalso inSet r1) eqs
+             val _ = Assert.tautologicalEquations "checkConstr' tautology checking" eqs
+           in
+             EqSpecInterBoundary H (constrs, specctx) boundaries
+           end
+       | Syn.IND_CONSTR_FUN (a,x,bx) =>
+           let
+             val w = Sym.new () (* is it possible to save this? *)
+             val goal = makeMem trace H (a, Syn.intoU (level, K.KAN))
+             val rest = checkConstr' (H @> (w, AJ.TRUE a), dimset) (constrs, specctx) (VarKit.rename (w, x) bx) level
```

Comment:
@ecavallo This is to say that, in the Gamma of the constructor data, everything still needs to be Kan. I believe it is still the case even with the 5-level type theory?